change: [M3-7383] - Replace hardcoded prices for LKE HA with data from `lke/types` API endpoint

carrillo-erik opened the pull request:
## Description 📝
The changes in this PR replace the hardcoded price values for LKE High Availability with the data from the `lke/types` endpoint.

## Changes  🔄
List any change relevant to the reviewer.
- Removal of the `LK_HA_PRICE` const.
- Replacement of outstanding `default imports` with `named imports`.
- `ClusterDetail` is renamed to `KubernetesClusterDetail`.
- New `kubernetesTypeFactory` type factory and corresponding tests.
- New DC Specific Pricing queries for Kubernetes (LKE HA).
- New `lke/types` API endpoint.

## How to test 🧪
- Go through the normal Kubernetes Create workflow.

### Verification steps
(How to verify changes)
- Verify that existing LKE create workflows behave as expected.
- Verify that switching between different regions changes the price. In particular `Jakarta, ID (id-cgk)` and `Sao Paulo, BR (br-gru)` should update the price for `LKE HA` from $60.00 to $72.00 and $84.00, respectively.
- Verify that the total price reflects the `LKE HA` selection based on the appropriate region.
- Verify there are no visual regressions as a result of these changes.

## As an Author I have considered 🤔
*Check all that apply*

- [x] 👀 Doing a self review
- [ ] ❔ Our [contribution guidelines](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md)
- [ ] 🤏 Splitting feature into small PRs
- [x] ➕ Adding a changeset
- [x] 🧪 Providing/Improving test coverage
- [ ] 🔐 Removing all sensitive information from the code and PR description
- [ ] 🚩 Using a feature flag to protect the release
- [x] 👣 Providing comprehensive reproduction steps
- [ ] 📑 Providing or updating our documentation
- [ ] 🕛 Scheduling a pair reviewing session
- [ ] 📱 Providing mobile support
- [ ] ♿  Providing accessibility support

carrillo-erik commented on packages/manager/src/factories/types.ts:
```diff
@@ -172,3 +172,37 @@ export const volumeTypeFactory = Factory.Sync.makeFactory<PriceType>({
 });
+
+export const kubernetesTypeFactory = Factory.Sync.makeFactory<PriceType>({
+  // {
```
@mjac0bs @abailly-akamai 
Could one of you guys please chime in on this change?

Looking at the Tiered Pricing API Spec, the `lke/types` endpoint returns more than one object. One for `LKE Standard Availability` and the other for `LKE High Availability`. We are only concerned with the latter for removal of the hardcoded values. However, if I try to create the factory with both objects I get an error due to the shape of the `PriceType` interface.

Is it okay to omit the `LKE Standard Availability` object here for the purpose of creating the factory?

mjac0bs replied:
This ([1](https://github.com/linode/manager/pull/10468/files#diff-c4904f695ece84849bc2372518313ff64475b5598ab35b45708266fa74e2e6e8R924-R930), [2](https://github.com/linode/manager/pull/10468/files#diff-eb6a0c856c00c5d1181386d59783fe18e78288417490a40da7ffd862e9a57db5) is what I did for `object-storage/types`, which also returns multiple price objects.

carrillo-erik replied:
@mjac0bs Great! I'll implement a similar pattern. Thanks for sharing.

carrillo-erik commented on packages/manager/src/features/Kubernetes/CreateCluster/HAControlPlane.tsx:
```diff
@@ -46,16 +53,29 @@ export const HAControlPlane = (props: HAControlPlaneProps) => {
       </FormLabel>
       <HACopy />
+      {hasHAPriceError && (
+        <Notice
```
@mjac0bs Disregard the previous comment. I figured out how to render a `<TextTooltip />` component within the label for the radio button. However, there is some refactoring required to solve some initial rendering issues with the `<TextTooltip />`. I also feel the logic could be simplified which in effect would take care of the rendering issue.

carrillo-erik commented on packages/manager/src/features/Kubernetes/KubernetesClusterDetail/UpgradeClusterDialog.tsx:
```diff
@@ -98,11 +110,8 @@ export const UpgradeKubernetesClusterToHADialog = (props: Props) => {
-          regionId: regionID,
-        })}{' '}
-        per month per cluster.
+        {/* TODO LKE/TYPES API: Show API error on UI */}
```
We should implement a similar solution to the one used in the `<HAControlPlace />` component to communicate to the user why HA pricing information is not available.

carrillo-erik commented on packages/manager/src/features/Kubernetes/KubernetesClusterDetail/UpgradeClusterDialog.tsx:
```diff
@@ -117,6 +126,7 @@ export const UpgradeKubernetesClusterToHADialog = (props: Props) => {
           </li>
         </ul>
       </Notice>
+      {/* TODO LKE/TYPES API: Disable in the event of API error */}
```
In addition to the comment above, I believe we should disable the upgrade option as we are specifically asking the user to agree to the terms of the new pricing contract. In the event where these details are not available due to API error, users cannot agree to terms they do not see in the UI.

carrillo-erik commented on packages/manager/src/features/Kubernetes/KubernetesClusterDetail/KubeClusterSpecs.tsx:
```diff
@@ -67,12 +70,15 @@ export const KubeClusterSpecs = (props: Props) => {
     : undefined;
 
+  {
+    /* TODO LKE/TYPES API: Show API error on UI */
```
This is another scenario where the pricing API can fail to load the specific DC price. In this view, we show the user the total price for running their LKE Cluster based on the configuration. If the pricing API fails, the current implementation would show the incorrect price. I believe we should show some type of indicator to the user in the UI that the price they see does not include the additional HA costs.

carrillo-erik commented:
#### Please Note:
Aside from resolving the merge conflicts, there's some optimizations that need to be implemented. There were additional use cases discovered and I'm waiting for input from UX on how to address them. For now, I've labeled this PR as `Work in Progress`.

abailly-akamai commented:
@carrillo-erik can we resolve the conflict before review?

mjac0bs commented on packages/manager/src/features/Kubernetes/CreateCluster/HAControlPlane.tsx:
```diff
@@ -46,17 +65,31 @@ export const HAControlPlane = (props: HAControlPlaneProps) => {
         <Typography variant="inherit">HA Control Plane</Typography>
       </FormLabel>
       <HACopy />
+      {isLoadingKubernetesTypes ? (
```
```suggestion
      {isLoadingKubernetesTypes && selectedRegionId !== '' ? (
```

The UX isn't quite right here. We'd only want to try to display the high availability price once a region has been selected.  Currently, we're seeing that loading spinner display without any region selection. `selectedRegionId` is initialized as an empty string instead of `undefined`, though I'm not sure why anymore. 

https://github.com/linode/manager/assets/114685994/ce648313-8c84-4957-8126-ed19f61b5818

It is slightly odd to have a loading spinner above the radio buttons in place of the notice while we determine whether we can get the price, but this is an edge case for slow networks or an unlikely erroring endpoint, so I think it's okay.

carrillo-erik replied:
Once again, I agree with you. The changes should reflect with the recent commit.

mjac0bs commented on packages/manager/src/features/Kubernetes/CreateCluster/HAControlPlane.tsx:
```diff
@@ -46,17 +65,31 @@ export const HAControlPlane = (props: HAControlPlaneProps) => {
+      ) : selectedRegionId && isErrorKubernetesTypes ? (
+        <Notice spacingBottom={4} spacingTop={24} variant="error">
+          <Typography>
+            The cost for HA Control Plane is not available at this time. Refer
```
```suggestion
            The cost for HA control plane is not available at this time. Refer
```

carrillo-erik replied:
Done.

mjac0bs commented on packages/manager/src/features/Kubernetes/CreateCluster/CreateCluster.tsx:
```diff
@@ -164,24 +168,23 @@ export const CreateCluster = () => {
+
+  const getHighAvailabilityPrice = getDCSpecificPriceByType({
+    regionId: selectedId ? selectedId : undefined,
+    type: kubernetesHighAvailabilityTypesData?.[1],
```
Rather than hardcoding with the index, I looked for the id that matched the object in the list that I wanted with object storage types, in case the API were to change at any point and return the types in a different order.

carrillo-erik replied:
I made the change here and the other places where the same logic was implemented.

mjac0bs reviewed (commented):
Here's some initial feedback, @carrillo-erik. Let me know if you have any questions, and thanks for working with Andrew on some of the UX aspects.

mjac0bs commented on packages/manager/src/queries/kubernetes.ts:
```diff
@@ -312,3 +315,22 @@ const getAllAPIEndpointsForCluster = (clusterId: number) =>
+    (results) => results.data
+  );
+
+export const typesQueries = createQueryKeys('types', {
+  lkeClusters: {
+    queryFn: getAllKubernetesTypes,
+    queryKey: null,
+  },
+});
```
Ha, this was timely - see also: https://github.com/linode/manager/pull/10556/files#diff-20a331beb2d3218a91fd88654aa14f348ed661f5d3d7b06d5ff723fe5ebd07d8R96-R99.

carrillo-erik commented on packages/manager/.changeset/pr-10505-changed-1716400104404.md:
```diff
@@ -0,0 +1,5 @@
+"@linode/manager": Changed
+---
+
+Replaced hardcoded values used in LKE HA pricing with data from API ([#10505](https://github.com/linode/manager/pull/10505))
```
Changed.

carrillo-erik commented on packages/manager/src/features/Kubernetes/CreateCluster/HAControlPlane.test.tsx:
```diff
@@ -18,9 +20,15 @@ describe('HAControlPlane', () => {
 
-  it('should not render an HA price when the price is undefined', () => {
+  it('should not render an HA price when there is a price error', () => {
+    const highAvailabilityPriceError = '--.--';
```
Removed and constant value imported.

carrillo-erik commented on packages/manager/src/features/Kubernetes/CreateCluster/HAControlPlane.test.tsx:
```diff
@@ -18,9 +20,15 @@ describe('HAControlPlane', () => {
-      <HAControlPlane {...props} highAvailabilityPrice={undefined} />
+      <HAControlPlane
+        {...props}
+        highAvailabilityPrice={highAvailabilityPriceError}
```
Same as comment above.

carrillo-erik commented on packages/manager/src/features/Kubernetes/CreateCluster/HAControlPlane.tsx:
```diff
@@ -46,17 +65,31 @@ export const HAControlPlane = (props: HAControlPlaneProps) => {
+            <Typography>
+              Yes, enable HA control plane.{' '}
+              {selectedRegionId
+                ? `For this region, HA Control Plane costs ($${highAvailabilityPrice}/month).`
```
I agree, the `( )`s don't make sense in a sentence.

carrillo-erik commented on packages/manager/src/features/Kubernetes/KubernetesClusterDetail/KubeClusterSpecs.tsx:
```diff
@@ -80,12 +96,31 @@ export const KubeClusterSpecs = (props: Props) => {
-      types: types ?? [],
-    }).toFixed(2)}/month`,
+    isLoadingKubernetesTypes ? (
+      <CircularProgress size={16} sx={{ marginTop: 2 }} />
```
Pulled and merged the latest from `develop` to this PR branch.

carrillo-erik commented on packages/manager/src/features/Kubernetes/CreateCluster/HAControlPlane.tsx:
```diff
@@ -46,17 +65,31 @@ export const HAControlPlane = (props: HAControlPlaneProps) => {
+        <Notice spacingBottom={4} spacingTop={24} variant="error">
+          <Typography>
+            The cost for HA Control Plane is not available at this time. Refer
+            to <Link to={getRegionPriceLink(selectedRegionId)}> pricing </Link>
```
Good eye!

mjac0bs commented on packages/manager/src/features/Kubernetes/CreateCluster/HAControlPlane.test.tsx:
```diff
@@ -18,9 +23,13 @@ describe('HAControlPlane', () => {
+  it('should not render an HA price when there is a price error', () => {
     const { queryAllByText } = renderWithTheme(
-      <HAControlPlane {...props} highAvailabilityPrice={undefined} />
+      <HAControlPlane
+        {...props}
+        highAvailabilityPrice={UNKNOWN_PRICE}
+        isErrorKubernetesTypes={true}
+      />
```
Can we assert that the error is displaying as expected here?  (Expect the notice text to be in the document, expect the unknown price ($--.--) to be in the document)

Currently, this just renders the component with an error.

mjac0bs commented on packages/manager/src/features/Kubernetes/KubeCheckoutBar/KubeCheckoutBar.test.tsx:
It would also be nice to have some test coverage here to confirm that if the HA price is not available, an unknown price will display and the total will be the sum of the node pools.

carrillo-erik replied:
Addressed both of your comments by updating the tests.

mjac0bs reviewed (approved):
Thanks for addressing feedback, @carrillo-erik! 

Confirmed that I'm seeing loading and error state when expected, and otherwise seeing HA pricing from the API reflected on the Kube create and details pages for normal regions, Jakarta, and Sao Paulo.

Left a few minor comments about some more clean up, but approving pending those are addressed.

abailly-akamai commented on packages/manager/src/features/Kubernetes/CreateCluster/CreateCluster.tsx:
```diff
@@ -58,6 +50,14 @@ import {
+
 export const CreateCluster = () => {
   const { classes } = useStyles();
   const [selectedRegionID, setSelectedRegionID] = React.useState<string>('');
```
nit: `const [selectedRegionId, setSelectedRegionId] = React.useState<string>('');`

abailly-akamai commented on packages/manager/src/features/Kubernetes/KubeCheckoutBar/KubeCheckoutBar.tsx:
```diff
@@ -22,13 +21,13 @@ import {
   hasAgreed: boolean;
   highAvailability?: boolean;
-  highAvailabilityPrice: number | undefined;
+  highAvailabilityPrice?: string;
```
You are removing a required prop here, which isn't needed. If your only change is the `number` > `string` type change, keep it required:

`highAvailabilityPrice: string | undefined;`

carrillo-erik replied:
Thanks for pointing this out.

abailly-akamai commented on packages/manager/src/features/Kubernetes/KubernetesClusterDetail/KubernetesClusterDetail.tsx:
```diff
@@ -15,34 +15,31 @@ import {
 import UpgradeKubernetesVersionBanner from './UpgradeKubernetesVersionBanner';
 
-export const KubernetesClusterDetail = () => {
+export const KubernetesClusterDetail = React.memo(() => {
```
you don't have props, there's nothing to memo here

carrillo-erik replied:
Removed.

abailly-akamai commented on packages/manager/src/features/Kubernetes/KubernetesClusterDetail/UpgradeClusterDialog.tsx:
```diff
@@ -70,6 +82,19 @@ export const UpgradeKubernetesClusterToHADialog = (props: Props) => {
+      : undefined;
+
+    return dcSpecificPrice ? parseFloat(dcSpecificPrice) : undefined;
+  };
```
You are repeating yourself with this function (also appears in CreateCluster). Make it a util and unit test it

carrillo-erik replied:
Part of my problem here was that I was trying to recycle old logic, however, you're feedback made me rethink this over and I refactored the implementation in all three components where this is used. Please let me know if you believe this can be improved.

abailly-akamai reviewed (approved):
Pr looks good overall, so does the UI. great job overall with coding conventions and attention to details providing fallbacks in the UI.

There's a few non blocking items to cleanup, approving pending mine & Mariah's comments

carrillo-erik commented on packages/manager/src/utilities/pricing/constants.ts:
```diff
@@ -16,6 +15,9 @@ export const UNKNOWN_PRICE = '--.--';
   'There was an error retrieving prices. Please reload and try again.';
+export const HA_UPGRADE_PRICE_ERROR_MESSAGE =
+  'Upgrading to HA is not available at this time. Try again later.';
+export const HA_PRICE_ERROR_MESSAGE = `The cost for HA Control Plane is not available at this time.`;
```
👍

carrillo-erik commented on packages/manager/src/features/Kubernetes/KubernetesClusterDetail/KubeClusterSpecs.tsx:
```diff
@@ -80,12 +100,31 @@ export const KubeClusterSpecs = (props: Props) => {
+      <CircleProgress size="sm" sx={{ marginTop: 2 }} />
+    ) : cluster.control_plane.high_availability && isErrorKubernetesTypes ? (
+      <>
+        (${UNKNOWN_PRICE}/month)
```
Done.

carrillo-erik commented on packages/manager/src/features/Kubernetes/KubernetesClusterDetail/KubeSummaryPanel.tsx:
```diff
@@ -156,7 +164,14 @@ export const KubeSummaryPanel = (props: Props) => {
+            }
+            cluster={cluster}
+            isErrorKubernetesTypes={isErrorKubernetesTypes}
+            isLoadingKubernetesTypes={isLoadingKubernetesTypes}
```
Moved the logic over to `KubeClusterSpecs` component.

carrillo-erik commented on packages/manager/src/factories/linodes.ts:
Updated the values to reflect current pricing.

carrillo-erik commented on packages/manager/src/features/Kubernetes/KubernetesClusterDetail/KubeClusterSpecs.tsx:
```diff
@@ -80,12 +100,31 @@ export const KubeClusterSpecs = (props: Props) => {
+    ) : cluster.control_plane.high_availability && isErrorKubernetesTypes ? (
+      <>
+        (${UNKNOWN_PRICE}/month)
+        <TooltipIcon
```
Fixed, the alignment. By taking a closer look at this I also noticed that mobile view would cause a regression, that's fixed now.

carrillo-erik commented:
@abailly-akamai @mjac0bs 
There's been a few updates to this PR and it continually got bigger (although the changes were minimal). I'm more than happy to schedule a pair review session before merging this PR.

mjac0bs reviewed (approved):
LKE pricing looks good after the final suggestions I've left, so approving pending those are addressed. We need to make sure we don't render `undefined` in the UI while the HA price is loading. (Toggling loading and error state on and off in the RQ dev tools is useful in testing this.)

I'm not seeing any regressions, but I didn't go to every single place that selectedRegionId now can be `undefined` when it would previously be `null` or an empty string to make sure that's handled as expected.

abailly-akamai commented:
@carrillo-erik you have to fix conflicts and address feedback before this gets out of date again

carrillo-erik commented on packages/manager/src/factories/linodes.ts:
```diff
@@ -177,18 +177,18 @@ export const linodeTypeFactory = Factory.Sync.makeFactory<LinodeType>({
       hourly: 0.018,
       id: 'id-cgk',
-      monthly: 12,
+      monthly: 14.4,
```
I reverted these values.

carrillo-erik commented on packages/manager/src/features/Kubernetes/CreateCluster/CreateCluster.tsx:
```diff
@@ -255,7 +259,12 @@ export const CreateCluster = () => {
               <HAControlPlane
-                highAvailabilityPrice={getHighAvailabilityPrice(selectedId)}
+                highAvailabilityPrice={
+                  isErrorKubernetesTypes ? UNKNOWN_PRICE : highAvailabilityPrice
```
I made these changes and verified the error and loading states.

carrillo-erik commented on packages/manager/src/features/Kubernetes/KubeCheckoutBar/KubeCheckoutBar.test.tsx:
```diff
@@ -71,25 +71,46 @@ describe('KubeCheckoutBar', () => {
 
-    // 5 node pools * 3 linodes per pool * 10 per linode * 20% increase for Jakarta
-    await findByText(/\$180\.00/);
+    // 5 node pools * 3 linodes per pool * 14.40 per linode * 20% increase for Jakarta
```
Thanks for the explanation. I updated the comments and added one more test.

carrillo-erik commented on packages/manager/src/features/Kubernetes/KubeCheckoutBar/KubeCheckoutBar.tsx:
```diff
@@ -22,15 +20,17 @@ import {
   hasAgreed: boolean;
   highAvailability?: boolean;
-  highAvailabilityPrice: number | undefined;
+  highAvailabilityPrice: string | undefined;
```
Done as stated in the other comment.

carrillo-erik commented on packages/manager/src/features/Kubernetes/CreateCluster/HAControlPlane.tsx:
```diff
@@ -14,7 +14,7 @@ export interface HAControlPlaneProps {
   highAvailabilityPrice?: string;
```
Done.

carrillo-erik commented:
> @carrillo-erik you have to fix conflicts and address feedback before this gets out of date again

@abailly-akamai I've pushed my recent changes addressing the feedback.
